presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1116,34 +1116,76 @@ protected Scope visitJoin(Join node, Optional<Scope> scope)
             Scope left = process(node.getLeft(), scope);
             Scope right = process(node.getRight(), isLateralRelation(node.getRight()) ? Optional.of(left) : scope);
 
-            Scope output = createAndAssignScope(node, scope, left.getRelationType().joinWith(right.getRelationType()));
-
-            if (node.getType() == Join.Type.CROSS || node.getType() == Join.Type.IMPLICIT) {
-                return output;
-            }
-
             if (criteria instanceof JoinUsing) {
-                // TODO: implement proper "using" semantics with respect to output columns
                 List<Identifier> columns = ((JoinUsing) criteria).getColumns();
 
-                List<Expression> expressions = new ArrayList<>();
+                List<Field> joinFields = new ArrayList<>();
+
+                List<Integer> leftJoinFields = new ArrayList<>();
+                List<Integer> rightJoinFields = new ArrayList<>();
+
+                Set<Identifier> seen = new HashSet<>();
                 for (Identifier column : columns) {
-                    Expression leftExpression = new Identifier(column.getValue());
-                    Expression rightExpression = new Identifier(column.getValue());
+                    if (seen.contains(column)) {
+                        throw new SemanticException(DUPLICATE_COLUMN_NAME, column, "Column '%s' appears multiple times in USING clause", column.getValue());
```

Comment:
Is it really a semantic exception? To me it does not matter if some one write `USING(a)` or `USING(a,a,a,a)`. It is still semantically the same as `USING(a)`. Notice it is legal to write `ORDER BY a,a,a,a`.

Reply:
I need to check the spec carefully. This is what postgreSQL does, so I chose to be more conservative.

Reply:
Specifically, what's not clear is what that would do to the schema of the joined table. Would

```
SELECT * FROM (VALUES 1, 'a') t(k, v1) JOIN (VALUES 1, 'b') u(a, v2) USING (a, a)
```

produce `1, 1, 'a', 'b'` or `1, 'a', 'b'`?

Reply:
If duplicates were allowed, only SELECT * would be valid, since trying to reference the column by name would result in an "ambiguous reference" error.